### PR TITLE
PKCE Always Spring Boot

### PR DESCRIPTION
**Pull request by dogeared**
This blog post has the following features:

- [x] A GitHub Repository with a polished README
- [x] A title that's approved by Dev Marketing
- [x] A URL that's approved by Dev Marketing
- [x] Rendered locally and confirmed that no Markdown typos exist

Live deploy: 

https://5e277aed0e60ca0008ab6391--okta-blog.netlify.com/blog/2020/01/23/pkce-oauth2-spring-boot

### CODE REVIEW

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 1]
```diff
@@ -0,0 +1,338 @@
+---
+layout: blog_post
+title: "Use PKCE with OAuth 2.0 and Spring Boot for Better Security"
+author: dogeared
+description: "PKCE guards against replay attacks with authorization codes, even for confidential clients"
```
Comment (mraible) [inline _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]:
```suggestion
description: "PKCE guards against replay attacks with authorization codes, even for confidential clients."
```

**Comment by mraible** [comment]
@dogeared It seems unfortunate that Heroku isn't mentioned in the title or description, but I suppose that's OK. Title scores a 62, so no complaints there. Maybe your should embed (or at least link to) your Heroku video?

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+```bash
+=== peaceful-citadel-41978 Config Vars
+LOGGING_LEVEL_ORG_SPRINGFRAMEWORK_WEB_CLIENT: DEBUG
+OKTA_ADMIN_EMAIL:                             109cd922-ab20-468e-a41d-01f986553087@heroku.okta.com
+OKTA_ADMIN_PASSWORD:                          A$1169b533-cfe9-466c-9355-07d7a922981a
+OKTA_CLIENT_ORGURL:                           https://dev-155005.okta.com
+OKTA_CLIENT_TOKEN:                            007pm-Knwql_WrDDJIEFa74tt5WuuXh564kdPZsgni
+OKTA_OAUTH2_CLIENT_ID_SPA:                    0oaztv4iSIaEcJRxS4x5
+OKTA_OAUTH2_CLIENT_ID_WEB:                    0oazsmh9j6TegkxYD4x5
+OKTA_OAUTH2_CLIENT_SECRET_WEB:                70nx3iG-mxzufsjYeb1-Gn5CUeo_n3slTpzBuliI
```
[the remark this answers is not in the document]
Comment (dogeared) [reply _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: already did! ;) Thanks.

**Comment by dogeared** [comment]
Heh - that's actually in there in one of the notes:

**NOTE:** Clicking the deploy to heroku button will allocate an Okta org for you, create an Okta OpenID Connect web application, deploy the example application and set all the environment variables for the application to run. If you want to know more about the Okta Heroku Add-On, you can see a video of it [here](https://www.youtube.com/watch?v=ZRBXgLTMvuY).

**Comment by dogeared** [comment]
I need to take your "here" advice, though

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+
+Last year, there were two important specification drafts submitted for review. These serve as the "updated guidance" for best practices when using OAuth 2.0. The first is [OAuth 2.0 Security Best Current Practice](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-13). This is general best practices for securing modern applications with OAuth 2.0. The second draft is [OAuth 2.0 for Browser-Based Apps](https://tools.ietf.org/html/draft-parecki-oauth-browser-based-apps-00). This is best practices specifically for web apps. In this post, I focus on the first draft with practical application for Spring Boot with Spring Security apps.
```
[the remark this answers is not in the document]
Comment (dogeared) [reply _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: fixed

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+Okta's own [Spring Boot Starter](https://github.com/okta/okta-spring-boot) makes it very easy to get started.
+
+## Get Started with Okta's Spring Boot Starter
+
+You can find the full source code this post [here](https://github.com/oktadeveloper/okta-spring-boot-oauth2-pkce-example) or head on over to [start.spring.io](https://start.spring.io) to quickly create a Spring Boot app with everything you need for a confidential client. The only starters you need are: Spring Web, Okta and [Thymeleaf](https://www.thymeleaf.org/). Okta automatically brings in Spring Security. Thymeleaf is used for html templates. The core of your `pom.xml` should look something like this:
+
+```
```
[the remark this answers is not in the document]
Comment (dogeared) [reply _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: fixed

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+	</dependency>
+	<dependency>
+		<groupId>com.okta.spring</groupId>
+		<artifactId>okta-spring-boot-starter</artifactId>
+		<version>1.3.0</version>
+	</dependency>
+</dependencies>
+```
+
+**Note:** The example code uses Java 11
```
[the remark this answers is not in the document]
Comment (dogeared) [reply _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: fixed

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+The first part of the `configure` method is standard Spring Security configuration. Here, we express that the home page (`/`) and anything under the static `img` folder does not require authentication. Any other path will require authentication.
+
+The second part of the `configure` method checks for the `okta.oauth2.pkce-always` environment variable and if set, configures Spring Security's `authorizationRequestResolver` with our `CustomAuthorizationRequestResolver`.
+
+With the security configuration in place, we can now ensure that the application uses PKCE for OAuth 2.0, thus enhancing the overall security of the application.
+
+## Good News Everyone!
+
+In the very near future, once [this](https://github.com/spring-projects/spring-security/pull/7804) pull request is merged and a new version of Spring Security is released (as well as the new version of the Spring Boot Spring Security Starter), you won't need to use the custom authorization request resolver and the security configuration as shown above. PKCE with confidential clients will be the default behavior.
```
[the remark this answers is not in the document]
Comment (dogeared) [reply _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: fixed

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+This is aligned when the current security best practices as oultined [here](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-13#section-3.1.1).
```
[the remark this answers is not in the document]
Comment (dogeared) [reply _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: fixed

[the remark this answers is not in the document]
Comment (dogeared) [reply _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: fixed

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+
+For more on OAuth 2.0 and OpenID Connect, I recommend these:
```
[the remark this answers is not in the document]
Comment (dogeared) [reply _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: fixed

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+
+* [OAuth 2.0 Java Guide: Secure Your App in 5 Minutes](/blog/2019/10/30/java-oauth2)
+* [Use Okta Token Hooks to Supercharge OpenID Connect](/blog/2019/12/23/extend-oidc-okta-token-hooks)
+* [An Illustrated Guide to OAuth and OpenID Connect](/blog/2019/10/21/illustrated-guide-to-oauth-and-oidc)
+* [What's going on with the OAuth 2.0 Implicit Flow?](https://www.youtube.com/watch?v=CHzERullHe8)
+* [OAuth 2.0 Access Tokens explained](https://www.youtube.com/watch?v=BNEoKexlmA4)
+* [OAuth 2.0 and OpenID Connect (in plain English)](https://www.youtube.com/watch?v=996OiexHze0&t=5s)
+
+For more on PKCE, I recommend these posts and videos:
```
[the remark this answers is not in the document]
Comment (dogeared) [reply _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: fixed

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md (hunk 1 above)
Comment (dogeared) [reply _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: fixed

**Comment by dogeared** [comment]
I updated with all your changes @mraible

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+		<groupId>org.springframework.boot</groupId>
+		<artifactId>spring-boot-starter-web</artifactId>
+	</dependency>
+	<dependency>
+		<groupId>com.okta.spring</groupId>
+		<artifactId>okta-spring-boot-starter</artifactId>
+		<version>1.3.0</version>
+	</dependency>
+</dependencies>
+```
```
Comment (mraible) [inline _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: tabs > spaces

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+
+**NOTE:** The example code uses Java 11
```
Comment (mraible) [inline _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: Add a period at the end of this sentence.

File: _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+}
+```
+
+**NOTE**: The `addPkceParameters` and `createHash` methods are borrowed from the existing `org.springframework.security.oauth2.client.web.DefaultOAuth2AuthorizationRequestResolver` found in the current version of the Spring Security source code.
+
+The `customAuthorizationRequest` method is where the action is. Assuming the `OAuth2AuthorizationRequest` parameter is not null, the code:
+
+1. grabs any existing attributes and additional parameters maps from the request
+1. adds the required pkce attributes and additional parameters to the existing maps for each
+1. builds and returns a new `OAuth2AuthorizationRequest` that includes the pkce attributes and additional parameters maps.
```
[the remark this answers is not in the document]
Comment (mraible) [reply _source/_posts/2020-01-23-pkce-oauth2-spring-boot.md]: This still needs to be fixed.
